python/loyalty_workflow.py

```diff
@@ -0,0 +1,135 @@
+import logging
+from datetime import timedelta
+
+from temporalio import activity, workflow
+from temporalio.common import WorkflowIDReusePolicy
+from temporalio.exceptions import WorkflowAlreadyStartedError
+from temporalio.workflow import ParentClosePolicy
+from temporalio.client import WorkflowHandle
+
+import email_strings
+from shared import *
+
+
+# Basic activity that logs and does string concatenation
+@activity.defn(name="SendEmail")
+async def send_email(body: str):
+    activity.logger.info("Sending email with contents %s" % body)
+
+
+# Basic workflow that logs and invokes an activity
+@workflow.defn
+class CustomerLoyaltyWorkflow:
+    def __init__(self):
+        self.customer = None
+
+    @workflow.run
+    async def run(self, customer: Customer) -> str:
+        self.customer = customer
+        workflow.logger.info("Running workflow with parameter %s" % customer)
+        info = workflow.info()
+
+        if not info.continued_run_id:
+            await workflow.execute_activity(
+                send_email,
+                email_strings.EMAIL_WELCOME.format(self.customer.status.name),
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+
+        await workflow.wait_condition(
+            lambda: not self.customer.account_active or info.get_current_history_length() > EVENT_HISTORY_THRESHOLD
```

Comment:
What if a guest invite signal is received and it hits this threshold? You will never do some of the guest logic. It is usually recommended to have signal handlers populate some kind of queue (can just be a list) that is processed inside the primary workflow function for this reason.

Reply:
Made signal handling a queue.